[16.0][ADD] stock_shipment_composer

AungKoKoLin1997 opened the pull request:
This module can compose custom outbound “shipments” by reshuffling stock moves before the delivery is validated.
Useful when you need to aggregate stock moves from multiple deliveries into one shipment. You can also plan multiple shipments for a stock move.

@qrtl QT5613

AungKoKoLin1997 committed:
[ADD] stock_shipment_composer

yostashiro commented on stock_shipment_composer/views/stock_shipment_composer_views.xml:
```diff
@@ -0,0 +1,332 @@
+                                    <field
+                                        name="product_uom_qty"
+                                        string="Demand"
+                                        optionl="hide"
```
```suggestion
                                        optional="hide"
```

yostashiro commented on stock_shipment_composer/models/stock_shipment_composer_line.py:
```diff
@@ -0,0 +1,37 @@
+    move_state = fields.Selection(related="move_id.state", string="Move Status")
+    remarks = fields.Text()
+    state = fields.Selection(related="composer_id.state", store=True)
+    reserved_enough = fields.Boolean(compute="_compute_reserved_enough")
```
Can we add location fields.
```suggestion
    reserved_enough = fields.Boolean(compute="_compute_reserved_enough")
    location_id = fields.Many2one(
        related="move_id.location_id", string="Source Location"
    )
    location_dest_id = fields.Many2one(
        related="move_id.location_dest_id", string="Destination Location"
    )
```

yostashiro commented on stock_shipment_composer/views/stock_shipment_composer_views.xml:
```diff
@@ -0,0 +1,332 @@
+                                        context="{'is_shipment_composer': True}"
+                                        help="The number in brackets represents the unallocated quantity to composers in 'In Progress' or 'Done' states."
+                                    />
+                                    <field name="picking_id" optional="show" />
```
```suggestion
                                    <field name="picking_id" optional="show" />
                                    <field
                                        name="location_id"
                                        string="From"
                                        optional="hide"
                                    />
                                    <field
                                        name="location_dest_id"
                                        string="To"
                                        optional="hide"
                                    />
```

yostashiro commented on stock_shipment_composer/models/stock_shipment_composer_line.py:
```diff
@@ -0,0 +1,37 @@
+    )
+    move_id = fields.Many2one("stock.move", required=True)
+    picking_id = fields.Many2one(related="move_id.picking_id", string="Transfer")
+    quantity = fields.Float(required=True)
```
Can you check if this causes an issue when a composer is created from the wizard?

```suggestion
    quantity = fields.Float(compute="_compute_quantity", store="True", readonly=False)
```

```.py
    @api.depends("move_id")
    def _compute_quantity(self):
        for rec in self:
            rec.quantity = rec.move_id.composer_unallocated_qty
```

AungKoKoLin1997 replied:
There is no issue when we use compute method.

yostashiro reviewed (commented):
Partial review.

yostashiro commented on stock_shipment_composer/models/stock_shipment_composer_line.py:
```diff
@@ -22,6 +22,13 @@ class StockShipmentComposerLine(models.Model):
     state = fields.Selection(related="composer_id.state", store=True)
     reserved_enough = fields.Boolean(compute="_compute_reserved_enough")
+    location_id = fields.Many2one(related="move_id.location_id", string="From")
+    location_dest_id = fields.Many2one(related="move_id.location_dest_id", string="To")
```
Please do as I suggested for the strings. The current ones are for labels and the official field descriptions should be something more explicit.

AungKoKoLin1997 replied:
Done!

AungKoKoLin1997 committed:
[FIX] stock_shipment_composer: apply move sort before reservation

action_assign() built a sorted recordset but discarded the result, so
moves were reserved in default order instead of by priority/deadline.
Assign the sorted result back to moves before _action_assign().

AungKoKoLin1997 committed:
[IMP] stock_shipment_composer: add summary and maintainers to manifest
